Doc/regions.md

```diff
@@ -0,0 +1,453 @@
+
+The fields of an immutable object may not be modified.
+Its reference count can be modified and it can be collected if it is no longer referenced.
+This reference count may be concurrently modified.
+
+To allow globals to be used in a concurrent setting, we have an explicit operation to promote make the globals immutable.
+```
+make_globals_immutable()
+```
+This must have been run to start the concurrency runtime and enable explicit region creation.
+
+
+## Project plan
+
+We need to break the development in to several phases.
+
+### Phase 1: Deeply immutable objects
+
+This requires
+
+*  A flag to represent if an object is deeply immutable.
+*  If the flag is set, then the object cannot be mutated.  (This needs to be enforced in the runtime.)
+*  Any object reachable from a deeply immutable object is also deeply immutable.
+*  A `freeze` operation that makes an object and its reachable graph of objects deeply immutable. 
+
+### Phase 2: `make_globals_immutable`
+
+This requires a function that makes all global objects deeply immutable.
+
+### Phase 3: Explicit regions
+
+This requires
+
+* Adding a field to the object header to represent the region.
+* A new object type to represent regions.
+* An operation to promote an object into an explicit region.
+* This requires tracking the number of references from local objects to an explicit region (as described above) called the "local reference count", LRC.
+
+### Phase 4: Write barrier
+
+This requires a write-barrier that ensures the explicit region is maintained.
+
+There are N cases to consider (that may overlap):
+
+* Changing a reference from a "local object" from an "explicit region", then we need to remove a reference from the LRC.
+* Changing a reference from a "local object" to an "explicit region", then we need to add a reference to the LRC.
+* Creating a reference from an "explicit region" to a "local object", then we need to promote the local object to the explicit region.  This may fail if the local object can reach a different explicit region. 
+
+### Phase 5: Send/consume
+
+Add an operation that represents sending a region to another thread.
+This is a placeholder operation that will be turned into proper concurrency support later in Phase 7.
+
+If the region still has a non-zero LRC, then we need to 
+
+* scan the current frame to find all references to the region, and remove them.
+* If there are still references, then scan all objects (from the cycle detector DLL) to find a reference, and report that as an error.
+
+### Phase 6: Memory management
+
+The existing Python cycle detector in 3.12 uses a doubly linked list effectively per interpreter.
+We need to make each region contain its own DLL, and then we can scan the DLL for cycle detection.
+
+### Phase 7: Concurrency
+
+Expose `when` and integrate with multiple interpreters.
+
+### Phase 8: More efficient region representation.
+
+The doubly linked list and the additional region pointer is too much state.
```

Comment:
This section is not very clear. Are you suggesting that instead of one N-bit pointer, you have a 16-bit index (for example) into a global region table, which in turn holds the pointers to the region objects?